In [697]:
%config Completer.use_jedi = False

## load pretrained and finetuned model 

In [698]:
from sqlalchemy import create_engine
import pymysql
import pandas as pd
import pickle
import torch
import json
from itertools import chain
from time import time
from collections import defaultdict
from tqdm import tqdm
from sentence_transformers import SentenceTransformer, util,LoggingHandler
import logging
from script.utils import Config
# logging.basicConfig(format='%(asctime)s - %(message)s',
#                     datefmt='%Y-%m-%d %H:%M:%S',
#                     level=logging.INFO)
config = Config(config_file='/data/yuchen/projects/sentence_embedding/config/process_pos.json')
model_sts = SentenceTransformer('distilbert-base-nli-mean-tokens')#config['model_path']
# model_save_path = 'script/output/new-pos-samples-pos1.3million-neg1.3million-distilbert-base-nli-mean-tokens-2021-03-11_14-08-37'
model_save_path_2 = 'script/output/new-raw-pos-samples-pos2.6million-neg1.3million-distilbert-base-nli-mean-tokens-2021-03-15_17-59-51'
model_save_path_3 = 'script/output/new-pos-samples-pos2million-neg2million-level3distilbert-base-nli-mean-tokens-2021-03-17_14-39-52'
model_finutuned_2 = SentenceTransformer(model_save_path_2)
model_finutuned_3 = SentenceTransformer(model_save_path_3)

## prepare test data

In [570]:
pos_train_data = pickle.load(open('data/pos_train_date_duplicate.pkl','rb'))
transformer_result=json.load(open('/data/yuchen/projects/sentence_embedding/data/test/transformer_result.json','r'))

In [571]:
def get_gold(index):
    a= [list(chain(*v)) for k,v in transformer_result[index].items()]
    return set(list(chain(*a)))

In [572]:
'init mysql connection'
db_connection_str = 'mysql+pymysql://shenjiawei:jiaweiDH$z048Kue2*34@cat-cluster.cluster-cvieeiq0uwtk.ap-southeast-1.rds.amazonaws.com:3306/ai_cat_ca_dev'
db_connection = create_engine(db_connection_str)
def get_lni_sentences(lnis):
    keyword_sql= "select l.lni, c.content from lni_unique as l inner join case_sentence as c on l.lni = c.lni where l.lni in " + '(' + ','.join(list(map(lambda x:'"' + x + '"',lnis))) + ')'
    result=pd.read_sql(keyword_sql, con=db_connection)
    lni_sentence = result.groupby('lni').agg(list)
    return lni_sentence

In [573]:
# reviewed test gold summary 
REVIEWED_POSTION = 5

In [649]:
test_lnis = list(transformer_result.keys())[:REVIEWED_POSTION]
lni_paths = {}
for lni in test_lnis:
    lni_paths[lni] = pos_train_data[pos_train_data.lni == lni].path.unique()
test_cases = get_lni_sentences(list(lni_paths.keys()))

In [650]:
test_cases['path'] = None
test_cases['gold'] = None
for index,row in test_cases.iterrows():
    row['path'] = list(lni_paths[row.name])
    row['gold'] = list(chain(*[list(chain(*v)) for k,v in transformer_result[row.name].items()]))

In [651]:
test_cases

content  \
lni                                                                               
5F16-93C1-FFMK-M0SJ-00000-00  [L.J. WENDEN PROV. CT. J., Reasons for Judgmen...   
5F8P-SFG1-JGPY-X039-00000-00  [N.J. SPIES J., ENDORSEMENT, Overview, 1The Ap...   
5GCD-JP81-DXHD-G3NS-00000-00  [M.A. DUNN J. (orally), REASONS FOR JUDGMENT, ...   
5T3H-Y051-F4GK-M0TD-00000-00  [A.B. CHRENEK PROV. CT. J., Decision, Introduc...   
5VB3-JF51-FCYK-20CM-00000-00  [N.J. GARSON J.A., Reasons for Judgment, The j...   

                                                                           path  \
lni                                                                               
5F16-93C1-FFMK-M0SJ-00000-00  [Criminal law - Criminal Code offences - Wilfu...   
5F8P-SFG1-JGPY-X039-00000-00  [Criminal law - Elements of the offence - Mens...   
5GCD-JP81-DXHD-G3NS-00000-00  [Criminal law - Sentencing - Particular sancti...   
5T3H-Y051-F4GK-M0TD-00000-00  [Family law - Custody and access - Considerati...   
5VB3-JF51-FCYK-20CM-00000-00  [Criminal law - Criminal Code offences - Sexua...   

                                                                           gold  
lni                                                                              
5F16-93C1-FFMK-M0SJ-00000-00  [After a trial, the accused was convicted of t...  
5F8P-SFG1-JGPY-X039-00000-00  [34For these reasons, the appeal from these tw...  
5GCD-JP81-DXHD-G3NS-00000-00  [Commission of offence for criminal organizati...  
5T3H-Y051-F4GK-M0TD-00000-00  [1This is an application and cross application...  
5VB3-JF51-FCYK-20CM-00000-00  [1 James Fisher, a former police officer, appe...

## test

In [666]:
def transformer_rank(gold_sents,all_sents,topk,model):
    #Compute embedding for both lists
    embeddings1 = model.encode(gold_sents, convert_to_tensor=True)
    embeddings2 = model.encode(all_sents, convert_to_tensor=True)
    #Compute cosine-similarits
    cosine_scores = util.pytorch_cos_sim(embeddings1, embeddings2)
    result = defaultdict(list)
    tensor, indice = torch.topk(cosine_scores,topk,largest=True)
    similar_sents = []
    for i in range(indice.shape[0]):
        # print("gold_sent{}----------------------------------------------------------------".format(i))
        # print(gold_sents[i])
        for j in range(indice.shape[1]):
            result[gold_sents[i]].append((all_sents[indice[i,j]],str(cosine_scores[i][indice[i,j]].detach().numpy()))) 
            similar_sents.append(all_sents[indice[i,j]])
        #     print("\n{} \n Score: {:.4f}".format(all_sents[indice[i,j]], cosine_scores[i][indice[i,j]]))
        # print("--------------------------------------------------------------------------")
    return similar_sents

In [599]:
indices =list(test_cases.index)
avg_recall = []
irs = {}
golds = {}
for ir_num in range(10,100,10):  
    recall = []
#     ir_num = 80
    for index,row in test_cases.iterrows():
    # index = '5T3H-Y051-F4GK-M0TD-00000-00'
        paths = list(map(lambda x : '-'.join(x.split('-')[:3]),row.path))
        ir = transformer_rank(paths,row['content'],ir_num,model_finutuned_3)
        gold = transformer_rank(list(row['gold']),row['content'],1,model_sts)
        recall.append(len(set(ir).intersection(set(gold))) / len(set(gold)))
        irs[row.name] = ir
        golds[row.name] = gold
    avg_recall.append((ir_num,sum(recall) / len(recall)))

In [600]:
avg_recall

[(10, 0.20803640803640805),
 (20, 0.33990453990453984),
 (30, 0.4025567025567026),
 (40, 0.4253487253487254),
 (50, 0.48002368002368),
 (60, 0.556976356976357),
 (70, 0.5717911717911718),
 (80, 0.6322455322455323),
 (90, 0.6652421652421652)]

In [638]:
index =  '5VB3-JF51-FCYK-20CM-00000-00'
sens1 = list(map(lambda x : '-'.join(x.split('-')[:3]),test_cases.path[index]))
sens1 = list(set(paths))
sens2 = test_cases.content[index]
# transformer_rank(sens1,sens2,model=model_finutuned_3,topk=10)

## export result

In [684]:
index = '5F8P-SFG1-JGPY-X039-00000-00'
content = get_lni_sentences([index])['content'][index]

In [693]:
list(map(lambda x: '-'.join(x.split('-')[:3]),pos_train_data[pos_train_data.lni==index].path.unique()))

['Criminal law - Elements of the offence - Mens rea ',
 'Criminal law - Procedure - Pleas (ss. 606 – 613 Cr.C.) ',
 'Criminal law - Procedure - Pleas (ss. 606 – 613 Cr.C.) ',
 'Criminal law - Appeals - Grounds (ss. 675(1), 676(1) and 830(1) Cr.C.)',
 'Criminal law - Appeals - Burden on appellant']

In [696]:
rr = {}
topk =80
path = list(map(lambda x: '-'.join(x.split('-')[:3]),pos_train_data[pos_train_data.lni==index].path.unique()))
sens = transformer_rank(path,content,model=model_finutuned_3,topk=topk)
for sen, score in sens:
    rr[sen] = score
[k for k,v in rr.items() if float(v)<0.6]

['He candidly admitted, however, he is not sure exactly when his brother stopped taking medication in 2008, and that he does not know whether or not his brother was taking medication when he was detained following his arrest on June 28, 2008.',
 '32Although I have difficulty in believing the Appellant when he deposed in June 2009 that he had no memory of these events; even if that is true, it does not add to the evidence concerning his state of mind at the material times.']

In [644]:
result = pd.DataFrame(columns=['lni','path','retrived','content'])
indices1 = ['5F8P-SFB1-FCK4-G557-00000-00']
for index in indices:
    paths = set()
    for path in list(test_cases.path[index]):
        path = '-'.join(path.split('-')[:3])
        if path not in paths:
            ir = transformer_rank([path],test_cases.content[index],10,model_finutuned_3)
            goldsens = test_cases.gold[index]
            result= result.append({'lni':index,'path':path,'retrived':ir,'content':goldsens},ignore_index=True)
        paths.add(path)

KeyError: '5F8P-SFB1-FCK4-G557-00000-00'

In [639]:
index  = 5
print(result.lni[index])
print(result.path[index])
result.retrived[index]

5GCD-JP81-DXHD-G3NS-00000-00
Criminal law - Controlled drugs and substances 


['Possession for purpose of trafficking',
 'An appropriate sentence for an offender like Mr. Kane who, for the purpose of monetary gain, has played a trusted and necessary role in a sophisticated, commercial level conspiracy to traffic in significant amounts of cocaine, would be in the range of four years imprisonment.',
 'The Newfoundland and Labrador Court of Appeal, while not altering the sentence, stated: " an appropriate sentence for the offender  who, for the purpose of monetary gain, has played a trusted and necessary role in a sophisticated, commercial level conspiracy to traffic in significant amounts of cocaine, would be in the range of 4 years imprisonment."',
 "Offenders were involved in a complex drug operation that involved moving large quantities of cocaine, ecstasy, marijuana and oxycodone between provinces using the postal system -- while on release from their first set of charges, police found drugs and drug paraphernalia in offender's residence -- neither had a prior

In [643]:
result.to_csv('result1.csv')

## test result
'distilbert-base-nli-mean-tokens'
'''
[(10, 0.01899134199134199),
 (20, 0.12609390609390608),
 (30, 0.2452947052947053),
 (40, 0.29591741591741594),
 (50, 0.305008325008325),
 (60, 0.3643489843489844),
 (70, 0.3954911754911755),
 (80, 0.43996669996669996),
 (90, 0.5063469863469864)]
'''

'''distilbert-base-nli-mean-tokens-finetuned on with 2 million old pos samples, 2 million neg samples
4 epoch, batch 32'''
'''
[(10, 0.01818181818181818),
 (20, 0.06213786213786214),
 (30, 0.14133200133200133),
 (40, 0.15671661671661669),
 (50, 0.22514818514818513),
 (60, 0.2559174159174159),
 (70, 0.28463536463536465),
 (80, 0.2937262737262737),
 (90, 0.3754911754911755)]
'''

'''distilbert-base-nli-mean-tokens-finetuned on 1.3 million new pos samples, 
1.3 million neg samples, 4 epoch, batch 64'''
'''
[(10, 0.18763236763236762),
 (20, 0.3266866466866467),
 (30, 0.41679653679653683),
 (40, 0.4612720612720612),
 (50, 0.5053746253746254),
 (60, 0.5738061938061938),
 (70, 0.5871395271395271),
 (80, 0.6157109557109557),
 (90, 0.6535198135198135)]
'''

'''distilbert-base-nli-mean-tokens-finetuned on 3 million old and new pos snetences, 
2 million neg samples, 6 epoch, batch 64'''
'''
[(10, 0.20822140822140822),
 (20, 0.29146779146779145),
 (30, 0.34283124283124283),
 (40, 0.4043697043697043),
 (50, 0.4624116624116624),
 (60, 0.5291597291597292),
 (70, 0.5953379953379954),
 (80, 0.6477596477596477),
 (90, 0.6551670551670552)]
'''

'''distilbert-base-nli-mean-tokens-finetuned on 2 million new pos samples,  
2 million neg samples, 4 epoch, batch 64''' 肉眼观察出来实际效果不是很好
'''
[(10, 0.20803640803640805),
 (20, 0.33990453990453984),
 (30, 0.4025567025567026),
 (40, 0.4253487253487254),
 (50, 0.48002368002368),
 (60, 0.556976356976357),
 (70, 0.5717911717911718),
 (80, 0.6322455322455323),
 (90, 0.6652421652421652)]
'''

## cited but related sentences in test cases

In [437]:
a = json.load(open('data/test/transformer_result_cited.json','r'))

In [438]:
indices = list(a.keys())[:3]

In [448]:
result_dic={}
for index in indices:
    result_dic[index] = [{d:[z[0] for z in v]} for d,v in a[index].items()]

In [504]:
result = pd.DataFrame(columns=['lni','summary','cited_case_sentence'])
for k,v in result_dic.items():
    row = {}
    for d in v:
        row['lni'] = k
        row['summary'] = list(d.keys())[0]
#         print(type(list(d.values())))
        row['cited_case_sentence'] = list(d.values())[0]
#         print(row)
        result = result.append(row,ignore_index=True)
#         print(list(d.keys())[0])
#     break

In [507]:
result.to_csv('cited_related_sens.csv')